File: epoxy-processor/src/main/java/com/airbnb/epoxy/GeneratedModelInfo.java
package com.airbnb.epoxy;

import com.squareup.javapoet.AnnotationSpec;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.ParameterSpec;
import com.squareup.javapoet.ParameterSpec.Builder;
import com.squareup.javapoet.TypeName;
import com.squareup.javapoet.TypeVariableName;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import javax.lang.model.element.AnnotationMirror;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.element.VariableElement;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;

import static com.airbnb.epoxy.Utils.buildEpoxyException;

abstract class GeneratedModelInfo {
  private static final String RESET_METHOD = "reset";
  protected static final String GENERATED_CLASS_NAME_SUFFIX = "_";

  protected TypeName superClassName;
  protected TypeElement superClassElement;
  protected TypeName parametrizedClassName;
  protected ClassName generatedClassName;
  protected TypeName boundObjectTypeName;
  protected boolean shouldGenerateModel;
  /**
   * If true, any layout classes that exist that are prefixed by the default layout are included in
   * the generated model as other layout options via a generated method for each alternate layout.
   */
  protected boolean includeOtherLayoutOptions;
  // TODO: (eli_hart 5/16/17) Sort attributes alphabetically so overloaded setters are together
  protected final List<AttributeInfo> attributeInfo = new ArrayList<>();
  protected final List<TypeVariableName> typeVariableNames = new ArrayList<>();
  protected final List<ConstructorInfo> constructors = new ArrayList<>();
  protected final Set<MethodInfo> methodsReturningClassType = new LinkedHashSet<>();
  protected final List<AttributeGroup> attributeGroups = new ArrayList<>();

  /**
   * Get information about methods returning class type of the original class so we can duplicate
   * them in the generated class for chaining purposes
   */
  protected void collectMethodsReturningClassType(TypeElement modelClass, Types typeUtils) {
    TypeElement clazz = modelClass;
    while (clazz.getSuperclass().getKind() != TypeKind.NONE) {
      for (Element subElement : clazz.getEnclosedElements()) {
        Set<Modifier> modifiers = subElement.getModifiers();
        if (subElement.getKind() == ElementKind.METHOD
            && !modifiers.contains(Modifier.PRIVATE)
            && !modifiers.contains(Modifier.FINAL)
            && !modifiers.contains(Modifier.STATIC)) {
          TypeMirror methodReturnType = ((ExecutableType) subElement.asType()).getReturnType();
          if (methodReturnType.equals(clazz.asType())
              || typeUtils.isSubtype(clazz.asType(), methodReturnType)) {
            ExecutableElement castedSubElement = ((ExecutableElement) subElement);
            List<? extends VariableElement> params = castedSubElement.getParameters();
            String methodName = subElement.getSimpleName().toString();
            if (methodName.equals(RESET_METHOD) && params.isEmpty()) {
              continue;
            }
            methodsReturningClassType.add(new MethodInfo(methodName, modifiers,
                buildParamSpecs(params), castedSubElement.isVarArgs()));
          }
        }
      }
      clazz = (TypeElement) typeUtils.asElement(clazz.getSuperclass());
    }
  }

  protected List<ParameterSpec> buildParamSpecs(List<? extends VariableElement> params) {
    List<ParameterSpec> result = new ArrayList<>();

    for (VariableElement param : params) {
      Builder builder = ParameterSpec.builder(TypeName.get(param.asType()),
          param.getSimpleName().toString());
      for (AnnotationMirror annotation : param.getAnnotationMirrors()) {
        builder.addAnnotation(AnnotationSpec.get(annotation));
      }
      result.add(builder.build());
    }

    return result;
  }

  void addAttribute(AttributeInfo attributeInfo) {
    addAttributes(Collections.singletonList(attributeInfo));
  }

  void addAttributes(Collection<AttributeInfo> attributesToAdd) {
    removeMethodIfDuplicatedBySetter(attributesToAdd);
    for (AttributeInfo info : attributesToAdd) {
      int existingIndex = attributeInfo.indexOf(info);
      if (existingIndex > -1) {
        // Don't allow duplicates.
        attributeInfo.set(existingIndex, info);
      } else {
        attributeInfo.add(info);
      }
    }
  }

  void addAttributeIfNotExists(AttributeInfo attributeToAdd) {
    if (!attributeInfo.contains(attributeToAdd)) {
      addAttribute(attributeToAdd);
    }
  }

  private void removeMethodIfDuplicatedBySetter(Collection<AttributeInfo> attributeInfos) {
    for (AttributeInfo attributeInfo : attributeInfos) {
      Iterator<MethodInfo> iterator = methodsReturningClassType.iterator();
      while (iterator.hasNext()) {
        MethodInfo methodInfo = iterator.next();
        if (methodInfo.name.equals(attributeInfo.getFieldName())
            // checking for overloads
            && methodInfo.params.size() == 1
            && methodInfo.params.get(0).type.equals(attributeInfo.getTypeName())) {
          iterator.remove();
        }
      }
    }
  }

  TypeElement getSuperClassElement() {
    return superClassElement;
  }

  TypeName getSuperClassName() {
    return superClassName;
  }

  List<ConstructorInfo> getConstructors() {
    return constructors;
  }

  Set<MethodInfo> getMethodsReturningClassType() {
    return methodsReturningClassType;
  }

  ClassName getGeneratedName() {
    return generatedClassName;
  }

  List<AttributeInfo> getAttributeInfo() {
    return attributeInfo;
  }

  boolean shouldGenerateModel() {
    return shouldGenerateModel;
  }

  Iterable<TypeVariableName> getTypeVariables() {
    return typeVariableNames;
  }

  TypeName getParameterizedGeneratedName() {
    return parametrizedClassName;
  }

  /**
   * Get the object type this model is typed with.
   */
  TypeName getModelType() {
    return boundObjectTypeName;
  }

  static class ConstructorInfo {
    final Set<Modifier> modifiers;
    final List<ParameterSpec> params;
    final boolean varargs;

    ConstructorInfo(Set<Modifier> modifiers, List<ParameterSpec> params, boolean varargs) {
      this.modifiers = modifiers;
      this.params = params;
      this.varargs = varargs;
    }
  }

  static class MethodInfo {
    final String name;
    final Set<Modifier> modifiers;
    final List<ParameterSpec> params;
    final boolean varargs;

    MethodInfo(String name, Set<Modifier> modifiers, List<ParameterSpec> params,
        boolean varargs) {
      this.name = name;
      this.modifiers = modifiers;
      this.params = params;
      this.varargs = varargs;
    }

    @Override
    public boolean equals(Object o) {
      if (this == o) {
        return true;
      }
      if (o == null || getClass() != o.getClass()) {
        return false;
      }

      MethodInfo that = (MethodInfo) o;

      if (varargs != that.varargs) {
        return false;
      }
      if (name != null ? !name.equals(that.name) : that.name != null) {
        return false;
      }
      if (modifiers != null ? !modifiers.equals(that.modifiers) : that.modifiers != null) {
        return false;
      }
      return params != null ? params.equals(that.params) : that.params == null;
    }

    @Override
    public int hashCode() {
      int result = name != null ? name.hashCode() : 0;
      result = 31 * result + (modifiers != null ? modifiers.hashCode() : 0);
      result = 31 * result + (params != null ? params.hashCode() : 0);
      result = 31 * result + (varargs ? 1 : 0);
      return result;
    }
  }

  @Override
  public String toString() {
    return "GeneratedModelInfo{"
        + "attributeInfo=" + attributeInfo
        + ", superClassName=" + superClassName
        + '}';
  }

  void addAttributeGroup(String groupName, List<AttributeInfo> attributes)
      throws EpoxyProcessorException {

    AttributeInfo defaultAttribute = null;
    for (AttributeInfo attribute : attributes) {
      if (attribute.isRequired() || attribute.codeToSetDefault.isEmpty()) {
        continue;
      }

      boolean hasSetExplicitDefault =
          defaultAttribute != null && defaultAttribute.codeToSetDefault.explicit != null;

      if (hasSetExplicitDefault && attribute.codeToSetDefault.explicit != null) {
        throw buildEpoxyException(
            "Only one default value can exist for a group of attributes: " + attributes);
      }

      // Have the one explicit default value in the group trump everything else.
      if (hasSetExplicitDefault) {
        continue;
      }

      // If only implicit
      // defaults exist, have a null default trump default primitives. This makes it so if there
      // is a nullable object and a primitive in a group, the default value will be to null out the
      // object.
      if (defaultAttribute == null
          || attribute.codeToSetDefault.explicit != null
          || attribute.hasSetNullability()) {
        defaultAttribute = attribute;
      }
    }

    AttributeGroup group = new AttributeGroup(groupName, attributes, defaultAttribute);
    attributeGroups.add(group);
    for (AttributeInfo attribute : attributes) {
      attribute.setAttributeGroup(group);
    }
  }

  static class AttributeGroup {
    final String name;
    final List<AttributeInfo> attributes;
    final boolean isRequired;
    final AttributeInfo defaultAttribute;

    AttributeGroup(String groupName, List<AttributeInfo> attributes,
        AttributeInfo defaultAttribute) throws EpoxyProcessorException {
      if (attributes.isEmpty()) {
        throw buildEpoxyException("Attributes cannot be empty");
      }

      if (defaultAttribute != null && defaultAttribute.codeToSetDefault.isEmpty()) {
        throw buildEpoxyException("Default attribute has no default code");
      }

      this.defaultAttribute = defaultAttribute;
      isRequired = defaultAttribute == null;
      this.name = groupName;
      this.attributes = new ArrayList<>(attributes);
    }

    CodeBlock codeToSetDefaultValue() {
      if (defaultAttribute == null || defaultAttribute.codeToSetDefault.isEmpty()) {
        throw new IllegalStateException("No default value exists");
      }

      return CodeBlock.of(defaultAttribute.setterCode(), defaultAttribute.codeToSetDefault.value());
    }
  }
}

File: epoxy-processor/src/main/java/com/airbnb/epoxy/processor/GeneratedModelInfo.kt
package com.airbnb.epoxy.processor

import androidx.room.compiler.processing.XElement
import androidx.room.compiler.processing.XTypeElement
import androidx.room.compiler.processing.XVariableElement
import com.airbnb.epoxy.ModelView
import com.airbnb.epoxy.processor.Utils.buildEpoxyException
import com.squareup.javapoet.AnnotationSpec
import com.squareup.javapoet.ClassName
import com.squareup.javapoet.ParameterSpec
import com.squareup.javapoet.TypeName
import com.squareup.javapoet.TypeVariableName
import java.util.ArrayList
import java.util.LinkedHashSet
import javax.lang.model.element.Modifier

abstract class GeneratedModelInfo(val memoizer: Memoizer) {

    lateinit var superClassElement: XTypeElement
    lateinit var superClassName: TypeName
    lateinit var parameterizedGeneratedName: TypeName
    lateinit var generatedName: ClassName

    /**
     * Get the object type this model is typed with.
     */
    var modelType: TypeName? = null
        protected set
    var shouldGenerateModel = false

    /**
     * If true, any layout classes that exist that are prefixed by the default layout are included in
     * the generated model as other layout options via a generated method for each alternate layout.
     */
    var includeOtherLayoutOptions = false

    val attributeInfo: MutableList<AttributeInfo> = mutableListOf()

    @get:Synchronized
    val attributeInfoImmutable: List<AttributeInfo>
        get() = attributeInfo.toList()

    val typeVariableNames: MutableList<TypeVariableName> = ArrayList()
    val constructors: MutableList<ConstructorInfo> = ArrayList()
    val methodsReturningClassType: MutableSet<MethodInfo> = LinkedHashSet()
    val attributeGroups: MutableList<AttributeGroup> = ArrayList()
    private val attributeToGroup = mutableMapOf<AttributeInfo, AttributeGroup>()
    val annotations: MutableList<AnnotationSpec> = ArrayList()

    /**
     * The info for the style builder if this is a model view annotated with @Styleable. Null
     * otherwise.
     */
    var styleBuilderInfo: ParisStyleAttributeInfo? = null
        private set

    /**
     * An option set via [com.airbnb.epoxy.ModelView.autoLayout] to have Epoxy create the
     * view programmatically
     * instead of via xml layout resource inflation.
     */
    var layoutParams = ModelView.Size.NONE

    /**
     * The elements that influence the generation of this model.
     * eg base model class for @EpoxyModelClass, view class for @ModelView, etc
     */
    fun originatingElements(): List<XElement> {
        return listOfNotNull(styleBuilderInfo?.styleBuilderElement)
            .plus(additionalOriginatingElements())
    }

    open fun additionalOriginatingElements(): List<XElement> = emptyList()

    /**
     * Get information about constructors of the original class so we can duplicate them in the
     * generated class and call through to super with the proper parameters
     */
    fun getClassConstructors(classElement: XTypeElement): List<ConstructorInfo> {
        return memoizer.getClassConstructors(classElement, memoizer)
    }

    /**
     * Get information about methods returning class type of the original class so we can duplicate
     * them in the generated class for chaining purposes
     */
    fun collectMethodsReturningClassType(superModelClass: XTypeElement) {
        methodsReturningClassType
            .addAll(memoizer.getMethodsReturningClassType(superModelClass.type, memoizer))
    }

    @Synchronized
    fun addAttribute(attributeInfo: AttributeInfo) {
        addAttributes(listOf(attributeInfo))
    }

    @Synchronized
    fun addAttributes(attributesToAdd: Collection<AttributeInfo>) {
        removeMethodIfDuplicatedBySetter(attributesToAdd)

        // Overwrite duplicates while preserving ordering
        for (attribute in attributesToAdd) {
            val existingIndex = attributeInfo.indexOf(attribute)
            if (existingIndex > -1) {
                attributeInfo[existingIndex] = attribute
            } else {
                attributeInfo.add(attribute)
            }
        }
    }

    @Synchronized
    fun addAttributeIfNotExists(attributeToAdd: AttributeInfo) {
        if (attributeToAdd !in attributeInfo) {
            addAttribute(attributeToAdd)
        }
    }

    private fun removeMethodIfDuplicatedBySetter(attributeInfos: Collection<AttributeInfo>) {
        for (attributeInfo in attributeInfos) {
            val iterator = methodsReturningClassType.iterator()
            while (iterator.hasNext()) {
                val (name, _, params) = iterator.next()
                if (name == attributeInfo.fieldName && params.size == 1 && params[0].type == attributeInfo.typeName) {
                    iterator.remove()
                }
            }
        }
    }

    val typeVariables: Iterable<TypeVariableName>
        get() = typeVariableNames

    val isStyleable: Boolean
        get() = styleBuilderInfo != null

    fun setStyleable(
        parisStyleAttributeInfo: ParisStyleAttributeInfo
    ) {
        styleBuilderInfo = parisStyleAttributeInfo
        addAttribute(parisStyleAttributeInfo)
    }

    val isProgrammaticView: Boolean
        get() = isStyleable || layoutParams != ModelView.Size.NONE

    fun hasEmptyConstructor(): Boolean {
        return constructors.isEmpty() || constructors.any { it.params.isEmpty() }
    }

    /**
     * @return True if the super class of this generated model is also extended from a generated
     * model.
     */
    val isSuperClassAlsoGenerated: Boolean
        get() = superClassElement.type.isSubTypeOf(memoizer.generatedModelType)

    data class ConstructorInfo internal constructor(
        val modifiers: Set<Modifier>,
        val params: List<ParameterSpec>,
        val varargs: Boolean
    )

    override fun toString(): String {
        return (
            "GeneratedModelInfo{" +
                "attributeInfo=" + attributeInfo +
                ", superClassName=" + superClassName +
                '}'
            )
    }

    @Throws(EpoxyProcessorException::class)
    fun addAttributeGroup(
        groupName: String?,
        attributes: List<AttributeInfo>
    ) {
        var defaultAttribute: AttributeInfo? = null
        for (attribute in attributes) {
            if (attribute.isRequired ||
                attribute.codeToSetDefault.isEmpty && !hasDefaultKotlinValue(
                        attribute
                    )
            ) {
                continue
            }
            val hasSetExplicitDefault =
                defaultAttribute != null && hasExplicitDefault(defaultAttribute)

            // Have the first explicit default value in the group trump everything else.
            // If there are multiple set just ignore the rest. This simplifies our lookup
            // of kotlin default params since it's hard to know exactly which function has
            // set a default param (if they have the same function name and param name)
            if (hasSetExplicitDefault) {
                continue
            }

            // If only implicit
            // defaults exist, have a null default trump default primitives. This makes it so if there
            // is a nullable object and a primitive in a group, the default value will be to null out the
            // object.
            if (defaultAttribute == null || hasExplicitDefault(attribute) ||
                attribute.hasSetNullability()
            ) {
                defaultAttribute = attribute
            }
        }
        val group = AttributeGroup(groupName, attributes, defaultAttribute)
        attributeGroups.add(group)
        attributes.forEach {
            attributeToGroup[it] = group
        }
    }

    /**
     * If this attribute is in a group, returns the other attributes contained in that group.
     */
    fun otherAttributesInGroup(attribute: AttributeInfo): List<AttributeInfo> {
        return attributeToGroup[attribute]
            ?.attributes
            ?.minus(attribute)
            ?: emptyList()
    }

    fun isOverload(attribute: AttributeInfo): Boolean {
        return attributeToGroup[attribute]?.attributes?.let { it.size > 1 } == true
    }

    fun attributeGroup(attribute: AttributeInfo): AttributeGroup? {
        return attributeToGroup[attribute]
    }

    class AttributeGroup internal constructor(
        groupName: String?,
        attributes: List<AttributeInfo>,
        defaultAttribute: AttributeInfo?
    ) {
        val name: String?
        val attributes: List<AttributeInfo>
        val isRequired: Boolean
        val defaultAttribute: AttributeInfo?

        init {
            if (attributes.isEmpty()) {
                throw buildEpoxyException("Attributes cannot be empty")
            }
            if (defaultAttribute != null && defaultAttribute.codeToSetDefault.isEmpty &&
                !hasDefaultKotlinValue(defaultAttribute)
            ) {
                throw buildEpoxyException("Default attribute has no default code")
            }
            this.defaultAttribute = defaultAttribute
            isRequired = defaultAttribute == null
            name = groupName
            this.attributes =
                ArrayList(attributes)
        }
    }

    companion object {
        const val RESET_METHOD = "reset"
        const val GENERATED_CLASS_NAME_SUFFIX = "_"
        const val GENERATED_MODEL_SUFFIX = "Model$GENERATED_CLASS_NAME_SUFFIX"

        fun buildParamSpecs(params: List<XVariableElement>, memoizer: Memoizer): List<ParameterSpec> {
            return params.map { it.toParameterSpec(memoizer) }
        }

        private fun hasDefaultKotlinValue(attribute: AttributeInfo): Boolean {
            return (attribute as? ViewAttributeInfo)?.hasDefaultKotlinValue == true
        }

        private fun hasExplicitDefault(attribute: AttributeInfo): Boolean {
            if (attribute.codeToSetDefault.explicit != null) {
                return true
            }

            return (attribute as? ViewAttributeInfo)?.hasDefaultKotlinValue == true
        }
    }
}
